### DISCUSSION TITLE
Support ordering by custom PostgreSQL function

### DISCUSSION THREAD

**Discussion by Aaron3219** [supabase/supabase#13849] [Questions] [answered]
_To be clear, this might be a bug instead of beeing unsupported_

Let's define a basic custom function `naturalsort` in Postgres:
```SQL
create or replace function naturalsort(text) returns bytea language sql immutable strict as $f$
select string_agg(
    convert_to(
      coalesce(
        r [2],
        length(length(r [1])::text) || length(r [1])::text || r [1]
      ),
      'SQL_ASCII'
    ),
    '\x00'
  )
from regexp_matches($1, '0*([0-9]+)|([^0-9]+)', 'g') r;
$f$;
```

When using this in an SQL query, I would use it like this:
```SQL
SELECT * FROM myTable ORDER BY naturalsort(myColumn) ASC;
```

I would expect this query to be equivalent in supabase-js with the following code:
```javascript
const { data, error } = await supabase
  .from('myTable')
  .select('*')
  .order('naturalsort(myColumn)', { ascending: true })
```

However, supabase-js returns an error:
```json
{
  "code": "PGRST104",
  "details": "Failed to parse naturalsort(myColumn)",
  "hint": null,
  "message": "Unexpected param or filter missing operator"
}
```

So if using custom SQL functions for ordering is unsupported, this would be a nice addition. If this is a bug, it should be fixed.

**Comment by steve-chavez** [comment, marked as answer]
You  can already do this with computed columns. You need to adapt your function like this:

```sql
create or replace function naturalsort(myTable) returns bytea language sql immutable strict as $f$
select string_agg(
    convert_to(
      coalesce(
        r [2],
        length(length(r [1])::text) || length(r [1])::text || r [1]
      ),
      'SQL_ASCII'
    ),
    '\x00'
  )
from regexp_matches($1.myColumn, '0*([0-9]+)|([^0-9]+)', 'g') r;
$f$;
```

Then:

```js
const { data, error } = await supabase
  .from('myTable')
  .select('*')
  .order('naturalsort', { ascending: true })
```

Note that you can use overloaded functions to apply `naturalsort` to other tables:

```sql
create or replace function naturalsort(anotherTable)..
create or replace function naturalsort(yetAnotherTable)..
```

Advantages of doing this.

- You whitelist the tables on which `naturalsort` can be applied
- Restrict your API surface(remember that supabase-js is meant to be called from the browser)

**Reply by Aaron3219** [reply]
So does that mean I would have to create a function for each column of my table?

Regardless, although this solution might be the best solution for the intended use case,
> remember that supabase-js is meant to be called from the browser

it adds an unnecessary layer of complexity for use cases which don't follow that principle for valid reasons.
